pip/pip-272.md

```diff
@@ -0,0 +1,214 @@
+
+1. `stateStorageProviderImplementation`: The implementation class for the state store which should implement the interface`StateStoreProvider`, such as `org.apache.pulsar.functions.instance.state.BKStateStoreProviderImpl`
+
+2. `stateStorageServiceUrl`: The service URL of state storage, such as: `bk://localhost:4181`
+
+`Runtime` and `RuntimeFactory`:
+
+Pulsar Function supports three kinds of runtime and, correspondingly, has three related `RuntimeFactory` to create them,
+
+1. ThreadRuntime: Each instance runs as a thread; it will load the `JavaInstanceRunnable`
+
+2. ProcessRuntime: Each instance runs as a process; it will start a `ProcessBuilder` with commands to start `JavaInstanceStarter`, the `JavaInstanceStarter` will load the `ThreadRuntime` and finally load the `JavaInstanceRunnable`, this is for `Java` function, for other languages, the commands will start related "instance"
+
+3. KubernetesRubtime: Function is submitted as Kubernetes StatefulSet by workers and each function instance runs as a pod; similar to the process runtime, the StatefulSet's command will start the `JavaInstanceStarter` and finally load the `JavaInstanceRunnable`, for other languages, the command will start related "instance"
+
+
+# Motivation
+
+<!--
+Describe the problem this proposal is trying to solve.
+
+* Explain what is the problem you're trying to solve - current situation.
+* This section is the "Why" of your proposal.
+-->
+
+Pulsar functions don't support configuring the `StateStoreProviderImplClass` with extra configurations, although there is a `Map<String, Object> config` parameter in the `StateStoreProvider#init` method, it's initialized as a map with only one field `stateStorageServiceUrl` in `javaInstanceRunnable`
+
+see:
+
+https://github.com/apache/pulsar/blob/master/pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java#L366-L369
+
+Then when using a custom `StateStoreProvider`, it cannot get other configurations from the `init` method.
+
+# Goals
+
+## In Scope
+
+<!--
+What this PIP intend to achieve once It's integrated into Pulsar.
+Why does it benefit Pulsar.
+-->
+
+Make `StateStoreProvider` "truly" configurable, custom `StateStoreProvider` can be configured more easily and explicitly.
+
+## Out of Scope
+
+<!--
+Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
+-->
+
+
+# High Level Design
+
+<!--
+Describe the design of your solution in *high level*.
+Describe the solution end to end, from a birds-eye view.
+Don't go into implementation details in this section.
+
+I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
+how you intend to solve it, end to end.
+
+DON'T
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+Make `StateStoreProvider` configurable using custom configurations.
```

Comment:
After reading the detailed design, I think we can provide a more detailed yet concise description of your solution:

We will add a configuration key to the Pulsar Function Worker configuration, containing the configuration for the `StateStoreProvider` implementation as a single-line JSON String. This JSON string, a `Map<String, Object>`, will be trickled down to different runtimes,  all the way to `JavaInstanceRunnable`, the ultimate Main class all runtimes use to execute the function. This class is also the one instantiating the `StateStoreProvider`.

Now that I actually described that, I suddenly realized something.
Say the operator decided to use a different `StateStoreProvider`.
Configuring it using a single-line JSON String would not be a convenient solution for the operator.

Shouldn't we have a mechanism to configure plugins with more than just a single property?

From what I've seen, Pulsar has more than 10 different types of plugins. I bet most of them need more than one config key. Today they put it plainly as multiple different keys on the same level one. Maybe they use a shared prefix in the name.

If we suddenly pivot to something else, it ruins the consistency and provides a low operator experience. 

@codelipenghui @merlimat - what do you think?

Reply:
Thanks for the improvements!

And as commented above, the `stateStorageConfig` is a `Map<String, Object>` rather than a JSON string.

For a custom `StateStoreProvider`, unlike other Pulsar plugins, what fields it needs for initialization is arbitrary, so we can only provide a `Map<String, Object>` to it and make the operator decide what fields to provide in the `functions_worker.yml`.

Reply:
Ok. I replied to this in the previous comment. Thanks